File: server/src/utils/bpParser.ts
interface OCRLocation {
  top: number;
  left: number;
  width: number;
  height: number;
}

interface OCRWord {
  words: string;
  location: OCRLocation;
}

interface BPResult {
  systolic?: number;
  diastolic?: number;
  heartRate?: number;
}

interface ParsedItem {
  value: number;
  location: OCRLocation;
  original: string;
}

export function parseBPData(wordsResult: OCRWord[]): BPResult {
  if (!wordsResult || wordsResult.length === 0) {
    return {};
  }

  // 1. Pre-process: Separate numbers and text
  const numbers: ParsedItem[] = [];
  const texts: { text: string; location: OCRLocation }[] = [];

  wordsResult.forEach(item => {
    const cleanText = item.words.replace(/[^\d.]/g, '');
    const val = parseFloat(cleanText);
    
    if (!isNaN(val) && cleanText.length > 0) {
      numbers.push({
        value: val,
        location: item.location,
        original: item.words
      });
    } else {
      texts.push({
        text: item.words.toUpperCase(),
        location: item.location
      });
    }
  });

  // 2. Filter numbers by reasonable range (loose filter)
  // BP monitors usually show 40-250. Heart rate 30-200.
  let validNumbers = numbers.filter(n => n.value >= 30 && n.value <= 300);

  // 3. Sort by vertical position (top to bottom)
  validNumbers.sort((a, b) => a.location.top - b.location.top);

  // 4. Strategy A: Keyword Anchoring (SYS, DIA, PULSE)
  // Try to find numbers near these keywords
  const resultA: BPResult = {};
  
  const findNearestNumber = (keywordPattern: RegExp, type: 'sys' | 'dia' | 'pul') => {
    const anchor = texts.find(t => keywordPattern.test(t.text));
    if (anchor) {
      // Find number that is below the anchor or to the right
      // "Below" means: number.top > anchor.top
      // "Close" means: vertical distance is small
      const candidate = validNumbers.find(n => {
        const verticalDist = n.location.top - anchor.location.top;
        // Must be below (-10 buffer) and within reasonable distance (e.g. 100px)
        return verticalDist > -10 && verticalDist < 150;
      });
      
      if (candidate) {
        if (type === 'sys') resultA.systolic = candidate.value;
        if (type === 'dia') resultA.diastolic = candidate.value;
        if (type === 'pul') resultA.heartRate = candidate.value;
        return true;
      }
    }
    return false;
  };

  findNearestNumber(/SYS|HIG|收缩/i, 'sys');
  findNearestNumber(/DIA|LOW|舒张/i, 'dia');
  findNearestNumber(/PUL|HR|RATE|心率|min/i, 'pul');

  // If Strategy A found everything, return it
  if (resultA.systolic && resultA.diastolic && resultA.heartRate) {
    return resultA;
  }

  // 5. Strategy B: Positional Inference (Fallback)
  // If we have 3 numbers vertically aligned, assume Sys -> Dia -> Pulse
  const resultB: BPResult = { ...resultA };
  
  // Remove numbers already used in Strategy A to avoid duplicates? 
  // Actually, let's just fill in the gaps.

  const unusedNumbers = validNumbers.filter(n => {
    return n.value !== resultB.systolic && 
           n.value !== resultB.diastolic && 
           n.value !== resultB.heartRate;
  });

  if (!resultB.systolic && unusedNumbers.length > 0) {
    // Take the top-most valid number as Systolic
    // Systolic is usually > 80
    const sysCandidate = unusedNumbers.find(n => n.value > 80);
    if (sysCandidate) {
      resultB.systolic = sysCandidate.value;
      // Remove from unused
      const idx = unusedNumbers.indexOf(sysCandidate);
      if (idx > -1) unusedNumbers.splice(idx, 1);
    }
  }

  if (!resultB.diastolic && unusedNumbers.length > 0) {
    // Take the next top-most as Diastolic
    // Diastolic is usually < Systolic
    const diaCandidate = unusedNumbers.find(n => {
      if (resultB.systolic) return n.value < resultB.systolic;
      return n.value < 130; // Guess
    });
    if (diaCandidate) {
      resultB.diastolic = diaCandidate.value;
      const idx = unusedNumbers.indexOf(diaCandidate);
      if (idx > -1) unusedNumbers.splice(idx, 1);
    }
  }

  if (!resultB.heartRate && unusedNumbers.length > 0) {
    // Remaining one is likely Heart Rate
    resultB.heartRate = unusedNumbers[0].value;
  }

  // 6. Final Sanity Check & Swap
  if (resultB.systolic && resultB.diastolic && resultB.systolic < resultB.diastolic) {
    const temp = resultB.systolic;
    resultB.systolic = resultB.diastolic;
    resultB.diastolic = temp;
  }

  return resultB;
}

File: server-go/utils/bp_parser.go
package utils

import (
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type BPResult struct {
	Systolic           *int    `json:"systolic,omitempty"`
	Diastolic          *int    `json:"diastolic,omitempty"`
	HeartRate          *int    `json:"heartRate,omitempty"`
	Confidence         float64 `json:"confidence"`
	ExtractionStrategy string  `json:"extractionStrategy"`
}

type parsedItem struct {
	Value    int
	Top      int
	Original string
}

func ParseBPData(wordsResult []OCRResult) BPResult {
	result := BPResult{}
	if len(wordsResult) == 0 {
		return result
	}

	var numbers []parsedItem
	var texts []struct {
		Text string
		Top  int
	}

	numRegex := regexp.MustCompile(`[^\d.]`)

	for _, item := range wordsResult {
		cleanText := numRegex.ReplaceAllString(item.Words, "")
		if val, err := strconv.ParseFloat(cleanText, 64); err == nil && cleanText != "" {
			numbers = append(numbers, parsedItem{
				Value:    int(val),
				Top:      item.Location.Top,
				Original: item.Words,
			})
		} else {
			texts = append(texts, struct {
				Text string
				Top  int
			}{
				Text: strings.ToUpper(item.Words),
				Top:  item.Location.Top,
			})
		}
	}

	// Filter by reasonable range (30-300)
	var validNumbers []parsedItem
	for _, n := range numbers {
		if n.Value >= 30 && n.Value <= 300 {
			validNumbers = append(validNumbers, n)
		}
	}

	// Sort by vertical position
	sort.Slice(validNumbers, func(i, j int) bool {
		return validNumbers[i].Top < validNumbers[j].Top
	})

	// Strategy A: Keyword anchoring
	sysPattern := regexp.MustCompile(`(?i)SYS|HIG|收缩`)
	diaPattern := regexp.MustCompile(`(?i)DIA|LOW|舒张`)
	pulPattern := regexp.MustCompile(`(?i)PUL|HR|RATE|心率|MIN`)

	findNearestNumber := func(pattern *regexp.Regexp) *int {
		for _, t := range texts {
			if pattern.MatchString(t.Text) {
				for _, n := range validNumbers {
					verticalDist := n.Top - t.Top
					if verticalDist > -10 && verticalDist < 150 {
						val := n.Value
						return &val
					}
				}
			}
		}
		return nil
	}

	result.Systolic = findNearestNumber(sysPattern)
	result.Diastolic = findNearestNumber(diaPattern)
	result.HeartRate = findNearestNumber(pulPattern)

	if result.Systolic != nil && result.Diastolic != nil && result.HeartRate != nil {
		result.ExtractionStrategy = "keyword"
		result.Confidence = 0.9
		return result
	}

	keywordPartial := result.Systolic != nil || result.Diastolic != nil || result.HeartRate != nil

	// Strategy B: Positional inference
	var unusedNumbers []parsedItem
	for _, n := range validNumbers {
		used := false
		if result.Systolic != nil && n.Value == *result.Systolic {
			used = true
		}
		if result.Diastolic != nil && n.Value == *result.Diastolic {
			used = true
		}
		if result.HeartRate != nil && n.Value == *result.HeartRate {
			used = true
		}
		if !used {
			unusedNumbers = append(unusedNumbers, n)
		}
	}

	if result.Systolic == nil && len(unusedNumbers) > 0 {
		for i, n := range unusedNumbers {
			if n.Value > 80 {
				val := n.Value
				result.Systolic = &val
				unusedNumbers = append(unusedNumbers[:i], unusedNumbers[i+1:]...)
				break
			}
		}
	}

	if result.Diastolic == nil && len(unusedNumbers) > 0 {
		for i, n := range unusedNumbers {
			if result.Systolic == nil || n.Value < *result.Systolic {
				val := n.Value
				result.Diastolic = &val
				unusedNumbers = append(unusedNumbers[:i], unusedNumbers[i+1:]...)
				break
			}
		}
	}

	if result.HeartRate == nil && len(unusedNumbers) > 0 {
		val := unusedNumbers[0].Value
		result.HeartRate = &val
	}

	// Sanity check: swap if systolic < diastolic
	if result.Systolic != nil && result.Diastolic != nil && *result.Systolic < *result.Diastolic {
		*result.Systolic, *result.Diastolic = *result.Diastolic, *result.Systolic
	}

	// Compute confidence score
	if keywordPartial {
		result.ExtractionStrategy = "keyword+positional"
		result.Confidence = 0.7
	} else {
		result.ExtractionStrategy = "positional"
		result.Confidence = 0.5
	}
	fields := 0
	if result.Systolic != nil {
		fields++
	}
	if result.Diastolic != nil {
		fields++
	}
	if result.HeartRate != nil {
		fields++
	}
	if fields < 2 {
		result.Confidence *= 0.5
	}

	return result
}
